### PR TITLE
Add custom version of Johnb43's Armor Modifier Ace

### PR DESCRIPTION
**Pull request by lambdatiger**
The mod adds a version of AMA
- Hit point armor minimum and maximum's set per side
- Unit specific hit point settings using `setVariable`
- API for setting both class default and specific unit armor values
- Some useful groupings of hit points such as "body", "limbs", "arms", and more (see fnc_resolveHitPoints)

Todo
- [x] Set defaults aimed at TvT damage
- [x] ~~Suppressed errors only for AMA~~ can't seem to pragma it away properly
- [x] MP testing (ty @TheCandianVendingMachine  & @AChesheireCat )

### CODE REVIEW

**Comment by lambdatiger** [comment]
Ready for review

File: addons/armor_modifier_ace/XEH_postInit.sqf [hunk 1]
```diff
@@ -0,0 +1,38 @@
+#include "script_component.hpp"
+
+// Call in post init, so that we are sure to overwrite default ACE damage handling and any other mods interfering with ACE medical
+["CAManBase", "initPost", {
+    params ["_unit"];
+
+    private _ehID = _unit getVariable ["ace_medical_HandleDamageEHID", -1];
```
Comment (TheCandianVendingMachine) [inline addons/armor_modifier_ace/XEH_postInit.sqf:7]: i mean... we could use `QACEGVAR`

File: addons/armor_modifier_ace/XEH_postInit.sqf [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+    _unit removeEventHandler ["HandleDamage", _ehID];
+
+    _ehID = _unit addEventHandler ["HandleDamage", {[_this] call FUNC(handleDamage)}];
+
+    _unit setVariable ["ace_medical_HandleDamageEHID", _ehID];
+}, true, [], true] call CBA_fnc_addClassEventHandler;
+
+[QGVAR(updateUnitArmor), {
+    params ["_unit", "_hitPoint", "_armorArray"];
+    if (isNull _unit || !alive _unit) exitWith {};
```
Comment (TheCandianVendingMachine) [inline addons/armor_modifier_ace/XEH_postInit.sqf:22]: can we not just use `!alive`?

File: addons/armor_modifier_ace/functions/fnc_handleDamage.sqf [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+    _unit setVariable ["ace_medical_lastDamageSource", _shooter];
+    _unit setVariable ["ace_medical_lastInstigator", _instigator];
+
+    ["ace_medical_woundReceived", [_unit, [[_newDamage, _hitPoint, _newDamage]], _shooter, "vehiclehit"]] call CBA_fnc_localEvent;
+
+    0
+};
+
+// Get setting for particular unit
+// I don't think we truely need 3 methods avaiable, I would rather choose two 2
```
Comment (TheCandianVendingMachine) [inline addons/armor_modifier_ace/functions/fnc_handleDamage.sqf:115]: is this comment from you or the source? if the former, what do you mean?

File: addons/armor_modifier_ace/functions/fnc_resolveHitPoints.sqf [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+ * Return Value:
+ * Array of hitpoints
+ *
+ * Public: Yes
+ */
+params [["_hitPoint", "", [""]],];
+
+_hitPoint = toLowerANSI _hitPoint;
+
+switch (toLowerANSI _hitPoint) do {
```
Comment (TheCandianVendingMachine) [inline addons/armor_modifier_ace/functions/fnc_resolveHitPoints.sqf:18]: duplicated?

File: addons/armor_modifier_ace/script_component.hpp [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    #define DEBUG_SETTINGS DEBUG_SETTINGS_ARMOR_MODIFIER_ACE
+#endif
+
+#include "\z\potato\addons\core\script_macros.hpp"
+#define GETMVAR(var1,var2) (missionNamespace getVariable [ARR_2(var1,var2)])
+#define SETMVAR(var1,var2,var3) (missionNamespace setVariable [ARR_3(var1,var2,var3)])
+
+#define DEFAULT_SETTINGS [ARR_3(1,0,0)]
+#define MINIMUM_SETTINGS [ARR_3(0.001,0,0)]
+#define DEFAULT_HASH_SETTINGS createHashMapFromArray [["hithead",[1,0,0]],["hitdiaphragm",[1,0,0]],["hitleftarm",[1,0,0]],["hitleftleg",[1,0,0]],["hitneck",[1,0,0]],["hitpelvis",[1,0,0]],["hitrightleg",[1,0,0]],["hitchest",[1,0,0]],["hitabdomen",[1,0,0]],["hitrightarm",[1,0,0]],["hitface",[1,0,0]]]
```
Comment (TheCandianVendingMachine) [inline addons/armor_modifier_ace/script_component.hpp:23]: Should these also be our defaults?

**Review by TheCandianVendingMachine — CHANGES_REQUESTED** [review]
a lot of questions, not necessarily needing changes but i would like clarification

File: addons/armor_modifier_ace/functions/fnc_handleDamage.sqf (hunk 3 above)
Comment (lambdatiger) [reply addons/armor_modifier_ace/functions/fnc_handleDamage.sqf:115]: Used to be 3 methods, now it's just 2. forgot to update it

File: addons/armor_modifier_ace/functions/fnc_resolveHitPoints.sqf (hunk 4 above)
Comment (lambdatiger) [reply addons/armor_modifier_ace/functions/fnc_resolveHitPoints.sqf:18]: yes

File: addons/armor_modifier_ace/script_component.hpp (hunk 5 above)
Comment (lambdatiger) [reply addons/armor_modifier_ace/script_component.hpp:23]: No, this is the default setting for not affecting anything

File: addons/armor_modifier_ace/XEH_postInit.sqf (hunk 2 above)
Comment (lambdatiger) [reply addons/armor_modifier_ace/XEH_postInit.sqf:22]: yes

**Review by TheCandianVendingMachine — APPROVED** [review]
did a once over, it looks good. performs in tests against players, so im assuming functionality is good

**Comment by lambdatiger** [comment]
Should removing this from implemented as part of this PR?
```sqf
// ACE Armor Adjuster
if (["AAA"] call ACEFUNC(common,isModLoaded)) then { _settings append [
["AAA_VAR_MOD_ENABLED", false]
];} else { _skip pushBack "AAA" };
```

**Review by PabstMirror — APPROVED** [review]

**Comment by PabstMirror** [comment]
we can do settings in another pr,
(I'm not sure if we're changing anything else like the truecon groaning)

**Pull request merged**